File: src-tauri/src/app/modules/repositories/fake/fake_repository.rs
use async_trait::async_trait;

use crate::app::modules::repositories::{
    interfaces::IRepository,
    models::{ModResult, VersionResult},
};

/// Fake repository used for testing.
/// Versions are keyed by Minecraft version.
#[derive(Default)]
pub struct FakeRepository {
    versions: Vec<VersionResult>,
}

impl FakeRepository {
    pub fn new() -> Self {
        Self { versions: vec![] }
    }

    pub fn with_versions(mut self, versions: Vec<VersionResult>) -> Self {
        self.versions = versions;
        self
    }
}

#[async_trait]
impl IRepository for FakeRepository {
    async fn search(&self, _query: &str, _page: usize) -> Vec<ModResult> {
        vec![]
    }

    async fn find(&self, _slug: &str) -> Option<ModResult> {
        None
    }

    async fn get_versions(
        &self,
        _project_id: &str,
        game_versions: &[String],
        _loaders: &[String],
    ) -> Vec<VersionResult> {
        self.versions
            .iter()
            .filter(|v| {
                v.minecraft_versions
                    .iter()
                    .any(|mc| game_versions.contains(mc))
            })
            .cloned()
            .collect()
    }
}

File: packages/core/src/testing/fake-repository.ts
import type { IRepository } from "../repositories/repository.interface.js";
import type { ModResult, VersionResult } from "../models/repository.js";
import type { ModFactory } from "./mod-factory.js";

/** Fake repository for testing — returns pre-configured versions, filters by MC version */
export class FakeRepository implements IRepository {
  private versions: VersionResult[] = [];

  withVersion(mf: ModFactory): this {
    this.versions.push(mf.toVersionResult());
    return this;
  }

  withVersions(versions: VersionResult[]): this {
    this.versions = versions;
    return this;
  }

  async search(_query: string, _page: number): Promise<ModResult[]> {
    return [];
  }

  async find(_slug: string): Promise<ModResult | undefined> {
    return undefined;
  }

  async getVersions(
    _projectId: string,
    gameVersions: string[],
    _loaders: string[],
  ): Promise<VersionResult[]> {
    return this.versions.filter((v) =>
      v.minecraftVersions.some((mc) => gameVersions.includes(mc)),
    );
  }
}
